### PR TITLE
Implementing methods: save() and load() 

### PR DESCRIPTION
**Pull request by h-haddad**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 1]
```diff
@@ -1413,23 +1414,141 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
    * Fetches record with given ID from database.
    * @exception VException      an exception may be raised by triggers
    */
-  fun fetchRecord(id: Int) {
-    TODO()
+  open fun fetchRecord(id: Int) {
+    val columns = getSearchColumns_()
+    val table = getSearchTables_()
+    val condition = Op.build { idColumn_ eq id  }
+    val query = table.slice(columns).select(condition)
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1528]: I think we might need to get other conditions from : VBlockDefaultOuterJoin.getFetchRecordCondition(fields)

Also move query inside the transaction block .

**Review by ZMarouani — CHANGES_REQUESTED** [review]

**Review by ZMarouani — APPROVED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 2]
```diff
@@ -1420,17 +1421,140 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+    }
+
+    // open database query, fetch tuples
+    val query = table.slice(columns!!).select(condition).orderBy(*orderBy.toTypedArray())
+
+    fetchCount = 0
+    transaction {
+      for (result in query) {
+        if (fetchCount < fetchSize) {
+          if (result[columns[1 + idqry]] == 0) {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1463]: use `columns[idqry]`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 3]
```diff
@@ -1420,17 +1421,140 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+            continue
+          }
+          fetchBuffer[fetchCount] = result[columns[1 + idqry]] as Int
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1466]: use `columns[idqry]`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 4]
```diff
@@ -1420,17 +1421,140 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+      if (fields[i].getColumnCount() > 0) {
+        idqry += 1
+      }
+    }
+
+    // open database query, fetch tuples
+    val query = table.slice(columns!!).select(condition).orderBy(*orderBy.toTypedArray())
+
+    fetchCount = 0
+    transaction {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1460]: remove `transaction`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 5]
```diff
@@ -1420,17 +1421,140 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+          fetchBuffer[fetchCount] = result[columns[1 + idqry]] as Int
+          if (fetchCount >= bufferSize) {
+            fetchCount += 1
+          } else {
+            var i = 0
+            var j = 0
+
+            while (i < fields.size) {
+              if (fields[i].getColumnCount() > 0) {
+                fields[i].setQuery_(fetchCount, query, 1 + j)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1475]: use `setQuery_(record: Int, query: ResultRow, column: Column<*>)`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 6]
```diff
@@ -964,6 +968,13 @@ abstract class VField protected constructor(width: Int, height: Int) : VConstant
     setObject(record, retrieveQuery_(query, column))
   }
 
+  /**
+   *
+   */
+  fun setQuery_(record: Int, query: org.jetbrains.exposed.sql.Query, column: Int) {
+    TODO()
+  }
+
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:977]: remove this

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 7]
```diff
@@ -1420,17 +1421,140 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+      if (!query.execute(this)!!.next()) {
+        /* Record does not exist anymore: it was deleted by another user */
+        throw VSkipRecordException()
+      } else {
+        /* set values */
+        var j = 0
+
+        for (field in fields) {
+          if (field.getColumnCount() > 0) {
+            field.setQuery_(query, columns[1 + j])
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1543]: use `setQuery_(query: ResultRow, column: Column<*>)`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 8]
```diff
@@ -950,6 +950,10 @@ abstract class VField protected constructor(width: Int, height: Int) : VConstant
     setQuery_(block!!.currentRecord, query, column)
   }
 
+  fun setQuery_(query: org.jetbrains.exposed.sql.Query, column: Column<*>) {
+    TODO()
+  }
+
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:956]: remove this method

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 9]
```diff
@@ -1420,17 +1421,140 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
    */
-  fun fetchRecord(id: Int) {
-    TODO()
+  fun fetchRecord(id: Int)  {
+    val columns = getSearchColumns()
+    val table = getSearchTables_()
+    val condition = mutableListOf<Op<Boolean>>()
+
+    condition.add(Op.build { idColumn eq id  })
+    condition.add(VBlockDefaultOuterJoin.getFetchRecordCondition(fields)!!)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1529]: Not null assertion here can cause NPE

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 10]
```diff
@@ -1420,17 +1421,141 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+    val condition = mutableListOf<Op<Boolean>>()
+
+    condition.add(Op.build { idColumn eq id  })
+    if (VBlockDefaultOuterJoin.getFetchRecordCondition(fields) != null) {
+      condition.add(VBlockDefaultOuterJoin.getFetchRecordCondition(fields)!!)
+    }
+
+    val query = table.slice(columns!!).select(condition.compoundAnd())
+
+    transaction {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1532]: no need for transaction in this context

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 11]
```diff
@@ -1420,17 +1421,141 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        for (field in fields) {
+          if (field.getColumnCount() > 0) {
+            for (result in query) {
+              field.setQuery_(result, columns[j])
+            }
+            j += 1
+          }
+        }
+        assert(!query.execute(this)!!.next()) { "too many rows" }
+      }
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1549]: use single() method, and catch `NoSuchElementException` and `IllegalArgumentException` seperately with corresponding error message

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 12]
```diff
@@ -1420,17 +1421,141 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+      if (!query.execute(this)!!.next()) {
+        /* Record does not exist anymore: it was deleted by another user */
+        throw VSkipRecordException()
+      } else {
+        /* set values */
+        var j = 0
+
+        for (field in fields) {
+          if (field.getColumnCount() > 0) {
+            for (result in query) {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1542]: why you're iterating over query?

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 13]
```diff
@@ -1420,17 +1421,141 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+
+    transaction {
+      if (!query.execute(this)!!.next()) {
+        /* Record does not exist anymore: it was deleted by another user */
+        throw VSkipRecordException()
+      } else {
+        /* set values */
+        var j = 0
+
+        for (field in fields) {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1540]: use `foreach`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 14]
```diff
@@ -1427,7 +1429,100 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        idqry += 1
+      }
+    }
+
+    // open database query, fetch tuples
+    var query = table!!.slice(columns!!).selectAll().orderBy(*orderBy.toTypedArray())
+
+    if (condition != null) {
+      query = table.slice(columns).select(condition).orderBy(*orderBy.toTypedArray())
+    }
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1469]: val query = if() ... else ...

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 15]
```diff
@@ -1427,7 +1429,100 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+
+    fetchCount = 0
+    for (result in query) {
+      if (fetchCount < fetchSize) {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1473]: have to stop when fetchCount < fetchSize is false

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 16]
```diff
@@ -1427,7 +1429,100 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+          fetchCount += 1
+        } else {
+          var j = 0
+
+          fields.forEach { field ->
+            if (field.getColumnCount() > 0) {
+              field.setQuery_(fetchCount, result, columns[j])
+              j += 1
+            }
+          }
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1488]: use foreachIndexed remove `j`

Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1488]: add empty line after

**Review by hfazai — COMMENTED** [review]
comments for load()

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 17]
```diff
@@ -1529,7 +1624,61 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+        VConstants.MOD_UPDATE -> updateRecord(0)
+        else -> throw InconsistencyException()
+      }
+    } else {
+      if (isIndexed()) {
+        /* first delete all deleted and changed old records */
+        for (i in 0 until bufferSize) {
+          if (isRecordFetched(i)) {
+            if (isRecordChanged(i)) {
+              transaction {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1645]: no need for transaction in this context

**Review by hfazai — COMMENTED** [review]
save()

**Pull request merged**